### PR TITLE
New version of ontid

### PR DESCRIPTION
**Pull request by siovanus**
1.add @context storage
2.add createTime and updateTime storage
3.add controller and access, etc in publickKey storage
4.add authentication storage
5.add service storage
6.add adn fix related api

### CODE REVIEW

File: cmd/abi/native_abi_script/ontid.json [hunk 1]
```diff
@@ -177,6 +217,239 @@
...
+    },
+    {
+      "name":"addAuthKey",
+      "parameters":[
+        {
+          "name":"id",
+          "type":"String"
+        },
+        {
+          "name":"ifNewPublicKey",
```
Comment (javajoker) [inline cmd/abi/native_abi_script/ontid.json]: isNewPublicKey?

File: smartcontract/service/native/ontid/group.go [hunk 2]
```diff
@@ -46,11 +47,14 @@ func (g *Group) Serialize() []byte {
 	for _, m := range g.Members {
 		switch t := m.(type) {
 		case []byte:
+			if !account.VerifyID(string(t)) {
+				panic("invalid ont id format")
```
Comment (javajoker) [inline smartcontract/service/native/ontid/group.go:51]: panic is forbidden in any native contract, there is no sandbox for it

File: smartcontract/service/native/ontid/group.go [hunk 3]
```diff
@@ -46,11 +47,14 @@ func (g *Group) Serialize() []byte {
...
+			}
 			sink.WriteVarBytes(t)
 		case *Group:
 			sink.WriteVarBytes(t.Serialize())
 		default:
-			panic("invlid member type")
+			panic("invalid member type")
```
Comment (javajoker) [inline smartcontract/service/native/ontid/group.go:57]: panic is forbidden in any native contract, there is no sandbox for it

File: smartcontract/service/native/ontid/owner.go [hunk 4]
```diff
@@ -90,40 +227,100 @@ func putAllPk(srvc *native.NativeService, key []byte, val []*owner) error {
...
+		publicKeys = append(publicKeys, &publicKey{pk, false, controller, a, authentication})
+		err = putAllPk_Version1(srvc, key, publicKeys)
+		if err != nil {
+			return 0, err
+		}
+		return uint32(size + 1), nil
+	}
+}
+
+func changePkAuthentication(srvc *native.NativeService, encId []byte, index uint32, authentication uint8, proof []byte) error {
```
Comment (javajoker) [inline smartcontract/service/native/ontid/owner.go:289]: shall verify proof if proof is provided and not empty

File: smartcontract/service/native/ontid/owner.go [hunk 5]
```diff
@@ -136,52 +333,203 @@ func findPk(srvc *native.NativeService, encID, pub []byte) (uint32, bool, error)
...
-				return index, errors.New("public key has already been revoked")
+	for i, v := range publicKeys {
+		if bytes.Equal(pub, v.key) && v.access != USE_ACCESS {
+			return uint32(i + 1), v.revoked, nil
+		}
+	}
+	return 0, false, nil
+}
+
+func revokeAuthKey(srvc *native.NativeService, encId []byte, index uint32, proof []byte) error {
```
Comment (javajoker) [inline smartcontract/service/native/ontid/owner.go:350]: shall verify proof if proof is provided and not empty

File: smartcontract/service/native/ontid/owner.go [hunk 6]
```diff
@@ -136,52 +333,203 @@ func findPk(srvc *native.NativeService, encID, pub []byte) (uint32, bool, error)
...
+	}
+	publicKeys[index-1].revoked = true
+	err = putAllPk_Version1(srvc, key, publicKeys)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func revokePk(srvc *native.NativeService, encId, pub, proof []byte) (uint32, error) {
```
Comment (javajoker) [inline smartcontract/service/native/ontid/owner.go:368]: shall verify proof if proof is provided and not empty

File: smartcontract/service/native/ontid/owner.go [hunk 7]
```diff
@@ -136,52 +333,203 @@ func findPk(srvc *native.NativeService, encID, pub []byte) (uint32, bool, error)
...
+		if err != nil {
+			return 0, err
+		}
+		return index, nil
 	}
-	if index == 0 {
-		return 0, errors.New("revoke failed, public key not found")
+}
+
+func revokePkByIndex(srvc *native.NativeService, encId []byte, index uint32, proof []byte) ([]byte, error) {
```
Comment (javajoker) [inline smartcontract/service/native/ontid/owner.go:419]: shall verify proof if proof is provided and not empty

File: smartcontract/service/native/ontid/owner.go [hunk 8]
```diff
@@ -136,52 +333,203 @@ func findPk(srvc *native.NativeService, encID, pub []byte) (uint32, bool, error)
...
+		return publicKeys[index].key, nil
 	}
-	putAllPk(srvc, key, owners)
-	return index, nil
 }
 
-func revokePkByIndex(srvc *native.NativeService, encID []byte, index uint32) ([]byte, error) {
-	key := append(encID, FIELD_PK)
-	owners, err := getAllPk(srvc, key)
+func setKeyAccessByIndex(srvc *native.NativeService, encId []byte, index uint32, access string, proof []byte) ([]byte, error) {
```
Comment (javajoker) [inline smartcontract/service/native/ontid/owner.go:460]: shall verify proof if proof is provided and not empty

File: smartcontract/service/native/ontid/authentication.go [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+	_, err := keypair.DeserializePublicKey(params.NewPublicKey.key)
+	if err != nil {
+		return utils.BYTE_FALSE, errors.New("add new auth key error: invalid key")
+	}
+	encId, err := encodeID(params.OntId)
+	if err != nil {
+		return utils.BYTE_FALSE, errors.New("add new auth key error: " + err.Error())
+	}
+
+	if checkIDState(srvc, encId) == flag_not_exist {
```
Comment (laizy) [inline smartcontract/service/native/ontid/authentication.go]: call `isValid` instead to handle ontid revoke case.

File: smartcontract/service/native/ontid/authentication.go [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+	params := new(SetAuthKeyParam)
+	if err := params.Deserialization(common.NewZeroCopySource(srvc.Input)); err != nil {
+		return utils.BYTE_FALSE, errors.New("set auth key error: deserialization params error, " + err.Error())
+	}
+	encId, err := encodeID(params.OntId)
+	if err != nil {
+		return utils.BYTE_FALSE, errors.New("set auth key error: " + err.Error())
+	}
+
+	if checkIDState(srvc, encId) == flag_not_exist {
```
Comment (laizy) [inline smartcontract/service/native/ontid/authentication.go]: ditto.

File: smartcontract/service/native/ontid/group.go (hunk 2 above)
Comment (laizy) [inline smartcontract/service/native/ontid/group.go]: move this function to `group_test.go`

Comment (siovanus) [reply smartcontract/service/native/ontid/group.go]: Done

**Review by laizy — APPROVED** [review]

**Pull request merged**